In [2]:
from datascience import *
import numpy as np
import matplotlib.pyplot as plt

## Data 8 - Project 1 Worksheet Demo

### Import Datasets

In [14]:
chocolates = Table.read_table('chocolate.csv')
ca = Table.read_table('ca.csv').select('State','Sex', 'Year','Name', 'Occurrence').take(np.arange(4))

### Question 1: Fun with Functions

**a)** The following code has a number of errors in it. Which ones can you identify?

```python
def hypotenuse(a, b)
```Returns the length of the hypotenuse of a right triangle, the squareroot of a squared + b squared```
squares = make_array(a, b)*2
sum = sum(squares)
squareroot = np.sqrt(sum)
print(squareroot)
	A = 5
	B = 5
	C = squareroot
```

In [ ]:
def hypotenuse(a, b)
```Returns the length of the hypotenuse of a right triangle, the squareroot of a squared + b squared```
squares = make_array(a, b)*2
sum = sum(squares)
squareroot = np.sqrt(sum)
print(squareroot)
    A = 5
    B = 5
    C = squareroot

**b)** Write a function that takes in one argument, a table tbl, another argument, a name of a column in that table col, and a boolean largest, and returns a table that contains the rows that have the ten largest or ten smallest values for the specified column, largest if the boolean largest is True, smallest if the boolean argument is False.

```python

	def top_ten(tbl, col, largest):
		sorted_tbl = _______________________________________
		ten_rows = _________________________________________
		return _____________________________________________
```

In [ ]:
def top_ten(tbl, col, largest):
    sorted_tbl = ...
    ten_rows = ...
    return ...

**c)** Can a function take no arguments? When would you use a function with no arguments? How do you call a function without arguments? How does that compare to using a function as an argument?


<details>
    <summary>Example</summary>
    <code>def function_no_arguments():
        return 'No Arguments'
function_no_arguments()</code>
</details>
<br>

In [ ]:
...

### Question 2

Ian has opened up a chocolate store where he sells small boxes of chocolates in groups of different sizes and colors. His table chocolates is as follows: 

In [10]:
chocolates

Color | Shape       | Amount | Price ($)
Dark  | Round       | 4      | 1.3
Milk  | Rectangular | 6      | 1.2
White | Rectangular | 12     | 2
Dark  | Round       | 7      | 1.75
Milk  | Rectangular | 9      | 1.4
Milk  | Round       | 2      | 1

Notice that the table contains multiple rows containing information about chocolates of the same color. We would like to figure out how many chocolates of each color he has for sale in total, and what the cost would be to purchase all chocolates of each unique color.

**a)** Write a line of code that will return a new table which displays the total number of boxes
for each color.

**b)** Write a line of code which will return a new table with the total number of chocolates and
the total cost for each unique color. For example, the row for “Dark” should have a total of 4+7=11 chocolates, and a total cost of \\$1.30 + \\$1.75 = 3.05.  


### Question 3

Some rows from the table ca are shown below. The table contains information about the most common baby names in California and the number of those occurrences in a particular year, from the years 1910-2019. (This dataset was submitted by a fellow Data 8 student!)

In [15]:
ca

State | Sex  | Year | Name     | Occurrence
CA    | F    | 1910 | Mary     | 295
CA    | F    | 1910 | Helen    | 239
CA    | F    | 1910 | Dorothy  | 220
CA    | F    | 1910 | Margaret | 163

**a)** Write a line of code that will return the most popular name over all the years. 
Hint: Think about how to use the second argument in .group 

**b)** Instead of the most popular name over all the years, write a line of code that will return the top 10 most popular names over all the years.

**c)** The top 10 names all appeared to be male names. Write a line of code that would return the most popular female names instead.

**d)** Write a line of code that will return the most popular female name in 1969

**e)** Write a function most_popular_female_name that takes in a year as an argument and returns the most popular female name in that year.

```python
def most_popular_female_name(year):
		_________________________________________________
```


In [ ]:
def most_popular_female_name(year):
    ...

**f)** The ca table is from 1920-2019. Define the years table with a column year and a row for each year from 1910-2019 (inclusive). Then create the table popular_female_names that has 2 columns, a year and a column for the female name that is most popular.

```python

years = ___________________________________________________________
most_popular_female_names_array = _________________________________
popular_female_names = ____________________________________________

```

In [ ]:
years = ...
most_popular_female_names_array = ...
popular_female_names = ...

**g)** Write a line of code that will generate the following bar chart: 

![barchart](barh.png)